File: api/schemas/vocabulary.py
"""Schémas Pydantic pour le vocabulaire Alteir et les guides narratifs."""
from typing import List, Optional, Dict, Any, Literal
from pydantic import BaseModel, Field
from datetime import datetime


class VocabularyTerm(BaseModel):
    """Terme du vocabulaire Alteir.
    
    Attributes:
        term: Le terme/concept.
        definition: Définition brève du terme.
        popularité: Niveau de popularité ("Mondialement", "Régionalement", "Localement", "Communautaire", "Occulte").
        category: Catégorie du terme ("Technologie", "Magie", "Race/Espèce", etc.).
        type: Type linguistique ("Calque linguistique", "Dérivation sémantique", etc.).
        origin: Origine du terme.
    """
    term: str = Field(..., description="Le terme/concept")
    definition: str = Field(default="", description="Définition brève du terme")
    popularité: str = Field(default="Occulte", description="Niveau de popularité")
    category: str = Field(default="Autre", description="Catégorie du terme")
    type: str = Field(default="", description="Type linguistique")
    origin: str = Field(default="", description="Origine du terme")


class VocabularyResponse(BaseModel):
    """Réponse contenant le vocabulaire filtré.
    
    Attributes:
        terms: Liste des termes filtrés.
        total: Nombre total de termes.
        filtered_count: Nombre de termes après filtrage.
        min_popularité: Niveau de popularité minimum utilisé pour le filtrage.
        statistics: Statistiques sur le vocabulaire.
    """
    terms: List[VocabularyTerm] = Field(default_factory=list, description="Liste des termes filtrés")
    total: int = Field(..., description="Nombre total de termes dans le vocabulaire")
    filtered_count: int = Field(..., description="Nombre de termes après filtrage")
    min_popularité: str = Field(..., description="Niveau de popularité minimum utilisé")
    statistics: Dict[str, Any] = Field(default_factory=dict, description="Statistiques sur le vocabulaire")


class VocabularySyncResponse(BaseModel):
    """Réponse de synchronisation du vocabulaire.
    
    Attributes:
        success: Indique si la synchronisation a réussi.
        terms_count: Nombre de termes synchronisés.
        last_sync: Timestamp de dernière synchronisation.
        error: Message d'erreur si la synchronisation a échoué.
    """
    success: bool = Field(..., description="Indique si la synchronisation a réussi")
    terms_count: int = Field(default=0, description="Nombre de termes synchronisés")
    last_sync: Optional[str] = Field(None, description="Timestamp de dernière synchronisation (ISO format)")
    error: Optional[str] = Field(None, description="Message d'erreur si échec")


class NarrativeGuideResponse(BaseModel):
    """Réponse contenant les guides narratifs.
    
    Attributes:
        dialogue_guide: Contenu du guide des dialogues.
        narrative_guide: Contenu du guide de narration.
        rules: Règles extraites (ton, structure, interdits, principes).
        last_sync: Timestamp de dernière synchronisation.
    """
    dialogue_guide: str = Field(default="", description="Contenu du guide des dialogues")
    narrative_guide: str = Field(default="", description="Contenu du guide de narration")
    rules: Dict[str, List[str]] = Field(default_factory=dict, description="Règles extraites par catégorie")
    last_sync: Optional[str] = Field(None, description="Timestamp de dernière synchronisation (ISO format)")


class NarrativeGuidesSyncResponse(BaseModel):
    """Réponse de synchronisation des guides narratifs.
    
    Attributes:
        success: Indique si la synchronisation a réussi.
        dialogue_guide_length: Longueur du guide des dialogues (caractères).
        narrative_guide_length: Longueur du guide de narration (caractères).
        last_sync: Timestamp de dernière synchronisation.
        error: Message d'erreur si la synchronisation a échoué.
    """
    success: bool = Field(..., description="Indique si la synchronisation a réussi")
    dialogue_guide_length: int = Field(default=0, description="Longueur du guide des dialogues")
    narrative_guide_length: int = Field(default=0, description="Longueur du guide de narration")
    last_sync: Optional[str] = Field(None, description="Timestamp de dernière synchronisation (ISO format)")
    error: Optional[str] = Field(None, description="Message d'erreur si échec")


class VocabularyStatsResponse(BaseModel):
    """Réponse contenant les statistiques du vocabulaire.
    
    Attributes:
        total: Nombre total de termes.
        by_popularité: Nombre de termes par niveau de popularité.
        by_category: Nombre de termes par catégorie.
        by_type: Nombre de termes par type linguistique.
    """
    total: int = Field(..., description="Nombre total de termes")
    by_popularité: Dict[str, int] = Field(default_factory=dict, description="Nombre par popularité")
    by_category: Dict[str, int] = Field(default_factory=dict, description="Nombre par catégorie")
    by_type: Dict[str, int] = Field(default_factory=dict, description="Nombre par type")


# Types de modes de sélection du vocabulaire
VocabularyMode = Literal["all", "auto", "none"]


class VocabularyConfig(BaseModel):
    """Configuration du vocabulaire par niveau de popularité.
    
    Permet de configurer indépendamment chaque niveau avec un mode de sélection :
    - "all": Inclure tous les termes de ce niveau
    - "auto": Inclure uniquement les termes de ce niveau qui sont mentionnés dans le contexte
    - "none": Ne pas inclure les termes de ce niveau
    
    Attributes:
        Mondialement: Mode de sélection pour le niveau Mondialement.
        Régionalement: Mode de sélection pour le niveau Régionalement.
        Localement: Mode de sélection pour le niveau Localement.
        Communautaire: Mode de sélection pour le niveau Communautaire.
        Occulte: Mode de sélection pour le niveau Occulte.
    """
    Mondialement: VocabularyMode = Field(default="all", description="Mode de sélection pour le niveau Mondialement")
    Régionalement: VocabularyMode = Field(default="all", description="Mode de sélection pour le niveau Régionalement")
    Localement: VocabularyMode = Field(default="none", description="Mode de sélection pour le niveau Localement")
    Communautaire: VocabularyMode = Field(default="none", description="Mode de sélection pour le niveau Communautaire")
    Occulte: VocabularyMode = Field(default="none", description="Mode de sélection pour le niveau Occulte")

File: frontend/src/api/vocabulary.ts
/**
 * API client pour le vocabulaire Alteir et les guides narratifs.
 */
import apiClient from './client'

export interface VocabularyTerm {
  term: string
  definition: string
  popularité: string
  category: string
  type: string
  origin: string
}

export interface VocabularyResponse {
  terms: VocabularyTerm[]
  total: number
  filtered_count: number
  min_popularité: string
  statistics: {
    by_popularité: Record<string, number>
    by_category: Record<string, number>
    by_type: Record<string, number>
  }
}

export interface VocabularySyncResponse {
  success: boolean
  terms_count: number
  last_sync: string | null
  error: string | null
}

export interface NarrativeGuideResponse {
  dialogue_guide: string
  narrative_guide: string
  rules: {
    ton: string[]
    structure: string[]
    interdits: string[]
    principes: string[]
  }
  last_sync: string | null
}

export interface NarrativeGuidesSyncResponse {
  success: boolean
  dialogue_guide_length: number
  narrative_guide_length: number
  last_sync: string | null
  error: string | null
}

export interface VocabularyStatsResponse {
  total: number
  by_popularité: Record<string, number>
  by_category: Record<string, number>
  by_type: Record<string, number>
}

/**
 * Récupère le vocabulaire Alteir, filtré par niveau de popularité.
 */
export async function getVocabulary(
  minPopularité?: string
): Promise<VocabularyResponse> {
  const params = minPopularité ? { min_popularité: minPopularité } : {}
  const response = await apiClient.get<VocabularyResponse>('/api/vocabulary', {
    params
  })
  return response.data
}

/**
 * Synchronise le vocabulaire depuis Notion.
 */
export async function syncVocabulary(): Promise<VocabularySyncResponse> {
  const response = await apiClient.post<VocabularySyncResponse>(
    '/api/vocabulary/sync'
  )
  return response.data
}

/**
 * Récupère les statistiques du vocabulaire.
 */
export async function getVocabularyStats(): Promise<VocabularyStatsResponse> {
  const response = await apiClient.get<VocabularyStatsResponse>(
    '/api/vocabulary/stats'
  )
  return response.data
}

/**
 * Récupère les guides narratifs.
 */
export async function getNarrativeGuides(): Promise<NarrativeGuideResponse> {
  const response = await apiClient.get<NarrativeGuideResponse>(
    '/api/narrative-guides'
  )
  return response.data
}

/**
 * Récupère uniquement les règles extraites des guides.
 */
export async function getExtractedRules(): Promise<{
  ton: string[]
  structure: string[]
  interdits: string[]
  principes: string[]
}> {
  const response = await apiClient.get<{
    ton: string[]
    structure: string[]
    interdits: string[]
    principes: string[]
  }>('/api/narrative-guides/rules')
  return response.data
}

/**
 * Synchronise les guides narratifs depuis Notion.
 */
export async function syncNarrativeGuides(): Promise<NarrativeGuidesSyncResponse> {
  const response = await apiClient.post<NarrativeGuidesSyncResponse>(
    '/api/narrative-guides/sync'
  )
  return response.data
}

